sql/src/main/java/io/crate/operation/projectors/BlockingSortingQueuedRowDownstream.java

```diff
@@ -0,0 +1,448 @@
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.operation.projectors;
+
+import com.google.common.collect.Ordering;
+import io.crate.core.collections.Row;
+import io.crate.core.collections.RowN;
+import io.crate.jobs.ExecutionState;
+import io.crate.operation.RowUpstream;
+import io.crate.operation.projectors.sorting.OrderingByPosition;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class BlockingSortingQueuedRowDownstream implements RowMerger {
+
+    private static final ESLogger LOGGER = Loggers.getLogger(BlockingSortingQueuedRowDownstream.class);
+
+    /**
+     * To prevent the Handles to block and unblock frequently when the queue size is reached
+     * we use two limits
+     *
+     * So the handle is blocked when queue has reached MAX_QUEUE_SIZE
+     * And unblock the handle if the queue has 0 rows
+     */
+    public static int MAX_QUEUE_SIZE = 1000;
+    public static int RESUME_AFTER = 10;
+
+    private final Ordering<Object[]> ordering;
+    private final List<BlockingSortingQueuedRowDownstreamHandle> downstreamHandles = new ArrayList<>();
+    private final List<RowUpstream> upstreams = new ArrayList<>();
+    private final AtomicInteger remainingUpstreams = new AtomicInteger(0);
+    private final AtomicBoolean downstreamAborted = new AtomicBoolean(false);
+    private final RowReceiver downstreamRowReceiver;
+    private final SharedArrayRef lowestToEmit;
+    private final int rowSize;
+    private final AtomicInteger runningHandles = new AtomicInteger(0);
+
+    private int pauseCount = 0;
+
+
+    public BlockingSortingQueuedRowDownstream(RowReceiver rowReceiver,
+                                              int rowSize,
+                                              int[] orderBy,
+                                              boolean[] reverseFlags,
+                                              Boolean[] nullsFirst) {
+        this.rowSize = rowSize;
+        List<Comparator<Object[]>> comparators = new ArrayList<>(orderBy.length);
+        for (int i = 0; i < orderBy.length; i++) {
+            comparators.add(OrderingByPosition.arrayOrdering(orderBy[i], reverseFlags[i], nullsFirst[i]));
+        }
+        ordering = Ordering.compound(comparators);
+        lowestToEmit = new SharedArrayRef(rowSize);
+        this.downstreamRowReceiver = rowReceiver;
+        rowReceiver.setUpstream(this);
+    }
+
+    @Override
+    public RowReceiver newRowReceiver() {
+        remainingUpstreams.incrementAndGet();
+        int handleIdx = runningHandles.getAndIncrement();
```

Comment:
is it really required to use 2 atomic integers here? looks like they are always adjusted synchronous..